Bond and GroupBond store bond order as float

goldmanm opened the pull request:
This pull request refactors RMG's Bond and GroupBond classes to represent bond order with a float instead of a string. This allows more flexibility with regards to representing bonds and reduces hard-coding of 'S', 'D', 'T', 'B' cases. In addition, I added methods to allow for back conversion if necessary.

This also modified methods that utilizes the order of molecule objects. The reading and writing of adjacency lists still involves string characters, so most of the database is intact and usable. The recipe actions of FORM_BOND, BREAK_BOND, and CHANGE_BOND now involve numbers instead of strings, so an [RMG-database pull request](https://github.com/ReactionMechanismGenerator/RMG-database/pull/156) involves the change of reaction recipes for each reaction family to address this problem. The code will not work without the RMG-database branch `bond_doubles` (which is in the pull request). 

This addresses issue #859 .

rwest commented:
@kehang - I think you have recently done a bunch of profiling of memory use (part of your pruning work). Would you like to check the impact of this change on memory use? I am intrigued to see if it makes a difference.  (If you can profile speed at the same time, ever better).

KEHANG commented:
@rwest sure! RMG-tests will automatically report memory usage and CPU time changes from master and this newly changed once Travis build passes. I will follow that.

goldmanm commented:
Note: I added a commit to allow using RMG-database branch `bond_doubles`. You might want to remove it before merging (now that it has passed). I also wrote how to make simultaneous pull requests of the database and code [in our wiki](https://github.com/ReactionMechanismGenerator/RMG-Py/wiki/simultaneous-update-of-database-and-RMG-Py). Feel free to update or fill in any information that was missing. Thanks to @nyee for showing me how to do it.

mliu49 commented on rmgpy/molecule/group.py:
```diff
@@ -542,9 +542,14 @@ class GroupBond(Edge):
     """
 
-    def __init__(self, atom1, atom2, order=None):
+    def __init__(self, atom1, atom2, order=[]):
```
I recall it being bad practice to set default values to mutable types, but I'm not sure how applicable that is in this particular case. It might be worth considering to switch back to None though.

goldmanm replied:
Thanks for bringing this up. I wasn't aware of this issue with Python.

mliu49 commented on rmgpy/molecule/group.py:
```diff
@@ -570,55 +575,98 @@ def copy(self):
-            else:
-                raise ActionError('Unable to update GroupBond due to CHANGE_BOND action: Invalid order "{0}".'.format(order))
+        for value in self.order:
+            newOrder.append(order + value)
```
Code suggestion (feel free to ignore):
```
newOrder = [value + order for value in self.order]
```

goldmanm replied:
great idea. I'll improve that.

mliu49 commented on rmgpy/molecule/group.py:
```diff
@@ -1392,7 +1440,7 @@ def classifyBenzeneCarbons(self, partners = None):
         for cbfAtom in cbfAtomList:
             for atom2, bond12 in cbfAtom.bonds.iteritems():
-                assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 'B'".format(self)
+                assert bond12.isBenzene(), "Cbf atom in {0} has a bond with an order other than 1.5".format(self)
```
The bond orders for Cbf atoms has always been a bit troublesome, but it's highlighted now because the bond orders are more explicit. When we made the changes to benzene bond treatment somewhat recently, we considered benzene bonds around a Cbf atom to have bond orders of 1.33. Switching to numerical bond orders makes this issue a bit more difficult to reconcile.

goldmanm replied:
The method used to find bond orders for atoms, getBondOrderForAtom, still counts the bonds around it as 1.33. The nearby Cb atom counts the bond as 1.5.

rwest replied:
....so whats the order of the bond stored as, if it's now stored as a float?

mliu49 commented on rmgpy/molecule/molecule.py:
```diff
@@ -434,20 +434,23 @@ class Bond(Edge):
     Attribute           Type                Description
     =================== =================== ====================================
-    `order`             ``str``             The :ref:`bond type <bond-types>`
+    `order`             ``short``             The :ref:`bond type <bond-types>`
```
float?

mliu49 commented on rmgpy/molecule/molecule.pxd:
```diff
@@ -81,12 +81,20 @@ cdef class Atom(Vertex):
+    
+    cpdef float getOrderNum(self)
+    
+    cpdef setOrderNum(self, short newOrder)
```
float?

mliu49 commented on rmgpy/molecule/molecule.py:
```diff
@@ -1667,4 +1710,4 @@ def getAromaticSSSR(self):
                     aromaticRings.append(ring0)
 
-        return aromaticRings
+        return aromaticRings
```
Could you remove spacing changes?

goldmanm commented:
@mliu49, Thanks for all the comments. I should have fixed everything easily fixable in your code. You did raise a good point about the bond orders around Cbf atoms. 

I looked up the pull request for Cbf #769, and it seems like the method that calculates the bond order, `getBondOrdersForAtom` I modified the method to ensure that functionality was conserved (see [line 484](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/bond_doubles/rmgpy/molecule/molecule.py)). Looking at that pull request for #769, I don't think any unit tests were written for `getBondOrdersForAtom`. That should definitely be done before it is merged.

goldmanm commented:
If the main method you're concerned about is `getBondOrdersForAtom` from the Atom class, then I wrote a test method that checks for the kekulized and aromatic versions of naphthalene and it calculates the number of bonds properly. 

Currently the code interprets bond order of 1.5 as benzene and does not include 1.33 as benzene. It doesn't account for the 1.33 anywhere except the `getBondOrdersForAtom` method.

rwest commented:
I've not checked if it's relevant but one thing to be careful of with this conversion to floating point bond orders is gotchas of this type:

```
In [1]: 0.2 + 0.1 == 0.3
Out[1]: False

In [2]: 0.2 + 0.1
Out[2]: 0.30000000000000004
```

KEHANG commented:
Good news from RMG-tests: for all the examples this branch generates same models as master does. Both the CPU time and memory consumption are almost same too. The tests were done against small examples usually finished within 5 mins, we might want to test big models to see the time and memory behaviors. But it's a good starting point.

goldmanm commented:
@rwest, you raised a good point about potential future bugs. I converted the methods, `isSingle`, `isDouble`, `isTriple`, and `isBenzene`, to compare within 10e-9 instead of equality to avoid any floating point issue. I also wrote a test method to test that it can handle floating point addition errors. If this doesn't cause a loss of efficiency, then it is definitely worth the robustness. I don't remember any code adding floating points together, though it could very well be used in the future.

@KEHANG, I think nick presented on RMG efficiency a while ago and mentioned different rate-limiting scenarios: generate-reactions, ODE solver, pruning, and a few other cases. I think this change likely would effect generate reactions (or another time when lots of molecule objects are created). It would be worth considering having a large 'generate reactions' rmg-test to emulate time & space efficiency on large models. Also, I added your commit text to the RMG-Wiki, so next time someone wants simultaneous changes to RMG-Py and the database, they can use that reference.

goldmanm commented:
i ran a 5 hour sulfur model with the current master and this branch to compare memory and output. The new bond system had memory differences and time differences within 0.5% of the master branch, which may even be within uncertainty of the nodes. 

One difference in output was observed. The [96th species](http://rmg.mit.edu/database/molecule/multiplicity%202%0A1%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B6%2CS%7D%20%7B9%2CS%7D%20%7B10%2CS%7D%0A2%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B11%2CS%7D%20%7B12%2CS%7D%20%7B13%2CS%7D%0A3%20%20C%20u0%20p0%20c0%20%7B5%2CS%7D%20%7B14%2CS%7D%20%7B15%2CS%7D%20%7B16%2CS%7D%0A4%20%20C%20u1%20p0%20c0%20%7B1%2CS%7D%20%7B2%2CS%7D%20%7B7%2CS%7D%0A5%20%20C%20u0%20p0%20c0%20%7B3%2CS%7D%20%7B7%2CD%7D%20%7B8%2CS%7D%0A6%20%20C%20u0%20p0%20c0%20%7B1%2CS%7D%20%7B8%2CD%7D%20%7B17%2CS%7D%0A7%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B5%2CD%7D%20%7B18%2CS%7D%0A8%20%20C%20u0%20p0%20c0%20%7B5%2CS%7D%20%7B6%2CD%7D%20%7B19%2CS%7D%0A9%20%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A10%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A11%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A12%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A13%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A14%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A15%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A16%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A17%20H%20u0%20p0%20c0%20%7B6%2CS%7D%0A18%20H%20u0%20p0%20c0%20%7B7%2CS%7D%0A19%20H%20u0%20p0%20c0%20%7B8%2CS%7D%0A) on this branch never appeared on the master branch. It seems to have been added right after the [first aromatic species](  http://rmg.mit.edu/database/molecule/1%20%20C%20u0%20p0%20c0%20%7B3%2CS%7D%20%7B9%2CS%7D%20%7B10%2CS%7D%20%7B11%2CS%7D%0A2%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B12%2CS%7D%20%7B13%2CS%7D%20%7B14%2CS%7D%0A3%20%20C%20u0%20p0%20c0%20%7B1%2CS%7D%20%7B5%2CB%7D%20%7B6%2CB%7D%0A4%20%20C%20u0%20p0%20c0%20%7B2%2CS%7D%20%7B5%2CB%7D%20%7B7%2CB%7D%0A5%20%20C%20u0%20p0%20c0%20%7B3%2CB%7D%20%7B4%2CB%7D%20%7B15%2CS%7D%0A6%20%20C%20u0%20p0%20c0%20%7B3%2CB%7D%20%7B8%2CB%7D%20%7B16%2CS%7D%0A7%20%20C%20u0%20p0%20c0%20%7B4%2CB%7D%20%7B8%2CB%7D%20%7B17%2CS%7D%0A8%20%20C%20u0%20p0%20c0%20%7B6%2CB%7D%20%7B7%2CB%7D%20%7B18%2CS%7D%0A9%20%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A10%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A11%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A12%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A13%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A14%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A15%20H%20u0%20p0%20c0%20%7B5%2CS%7D%0A16%20H%20u0%20p0%20c0%20%7B6%2CS%7D%0A17%20H%20u0%20p0%20c0%20%7B7%2CS%7D%0A18%20H%20u0%20p0%20c0%20%7B8%2CS%7D%0A). The new molecule was formed from disproportionation reactions of the added aromatic compound. After this extra one was added, both models added the same species to the core, including the similar [isomer of the first molecule](http://rmg.mit.edu/database/molecule/multiplicity%202%0A1%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B5%2CS%7D%20%7B9%2CS%7D%20%7B10%2CS%7D%0A2%20%20C%20u0%20p0%20c0%20%7B4%2CS%7D%20%7B11%2CS%7D%20%7B12%2CS%7D%20%7B13%2CS%7D%0A3%20%20C%20u0%20p0%20c0%20%7B5%2CS%7D%20%7B14%2CS%7D%20%7B15%2CS%7D%20%7B16%2CS%7D%0A4%20%20C%20u0%20p0%20c0%20%7B1%2CS%7D%20%7B2%2CS%7D%20%7B6%2CD%7D%0A5%20%20C%20u0%20p0%20c0%20%7B1%2CS%7D%20%7B3%2CS%7D%20%7B7%2CD%7D%0A6%20%20C%20u0%20p0%20c0%20%7B4%2CD%7D%20%7B8%2CS%7D%20%7B18%2CS%7D%0A7%20%20C%20u0%20p0%20c0%20%7B5%2CD%7D%20%7B8%2CS%7D%20%7B19%2CS%7D%0A8%20%20C%20u1%20p0%20c0%20%7B6%2CS%7D%20%7B7%2CS%7D%20%7B17%2CS%7D%0A9%20%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A10%20H%20u0%20p0%20c0%20%7B1%2CS%7D%0A11%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A12%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A13%20H%20u0%20p0%20c0%20%7B2%2CS%7D%0A14%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A15%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A16%20H%20u0%20p0%20c0%20%7B3%2CS%7D%0A17%20H%20u0%20p0%20c0%20%7B8%2CS%7D%0A18%20H%20u0%20p0%20c0%20%7B6%2CS%7D%0A19%20H%20u0%20p0%20c0%20%7B7%2CS%7D%0A). In the end, the new branch had one more species, which was the one discussed above. Since the models are so similar, I think this may just be from numerical error in the flux analysis, as it doesn't seem to impact model growth after it was added.

rwest commented:
I know 02711cfd238a25ada546410708451a628d1e0a2f was added to address a warning I made above about floating point arithmetic, but I'm not sure it's solves much. True it *would* detect if there was a bond for some reason stored with an order of 1.000000001  ensuring that `.isSingle()` would return True, but this is an unlikely scenario and may be redundant. If someone sets a bond order to 1.0 then `order==1.0` will be True anyway.   Furthermore, it would *not* prevent someone writing code like `if sum([bond.order for bond in molecule.bonds]) == 4.0`, in which they do not make any calls to the `.isSingle()`, `.isDouble()` methods etc. that you added the protection to. Protecting against future errors is hard though - so maybe we just check we're not currently making such errors, and put a warning somewhere in a docstring. 

My other question, which brought me back to this thread, is wondering if anyone has thoughts on how we'd do a van der Waals bond in this floating point representation? Something I'm about to add to our heterogeneous catalysis branch....  In the string-based method we're going with 'vdW', but what should the float be?  Would `0` cause any problems?  (be confused with `None` somewhere?).

And I didn't really understand how the fused benzene bonds are being dealt with (1.33333 or 1.5?) (or does getBondOrderForAtom return something different than the bond order?). Sorry I've not been paying close enough attention to know this, which has probably been explained somewhere.

goldmanm commented:
@rwest, Thanks for the concerns. Its good we ask these questions now, as opposed to later. I hope this clarifies some of the concerns. My main goal of this pull request is to refactor the code and not change functionality. 

**about potential floating point errors** - Like you, I also can't think of a reason why having a more flexible `isSingle` could be useful during a standard RMG run. It doesn't seem to add to computation time significantly. The example posed `if sum([bond.order for bond in molecule.bonds]) == 4.0` may not be the best way to check order if you expect floating point errors. I searched through the RMG-Py code for '.order' and re-factored any occurrence. The two occurrences I noticed was `transformLonePairs` in `solvation.py` where bonds were summed separately from `getBondOrderForAtom` and in `getBondOrderForAtom` which doesn't round the returned results (though it also didn't round before this pull request either). 

**van der waals** - I don't see a bond order of 0 causing any issues. I can't recall using it in boolean operations where it would be construed as False, since they have never been zero before. Though maybe I don't completely understand the causes of a `None` error. If there are restrictions on the number of bond objects tied to an atom, that would cause a problem in both the previous and new branch. 

**fused benzene bonds** - All benzene bonds have order of 1.5. The method performs just like before with the `getBondOrderForAtom`, since the main point of the PR is refactoring (so no functionality changes). I wrote a test (none existed before this PR), to ensure the constant functionality: `testGetBondOrdersForAtom` in MoleculeTest.py. This implementation of  `getBondOrderForAtom` seemed a little odd to me since the numeric order given for a bond depends on the other number of benzene bonds around that atom object. For example, a regular benzene bond is given the order of 1.3333 if this method is called on an atom with 2 other benzene bonds, but the same bond is given an order of 1.5 if called on an atom with only one other benzene bond. Even though the method seems odd in its implementation, I just made it function the same way as before since this PR isn't meant to fix these other issues.

goldmanm commented:
I added one more method isOrder() (plus unit tests) because @KEHANG recommended that it would be good to have a more general comparison for any future bond orders.

I was looking more into floating point errors. Since all our current bond values which are stored and reacted (0, 0.5, 1, 1.5, 2, 2.5) are able to be represented by less than 10 bits, we likely will not encounter any floating point errors in the future.  This would include @rwest's `VdW` and @mliu49's `B+` and `B-` bond types that haven't been merged. I wrote the code so that it should be able to take into account other bonds, that we may add in the future. 

@rwest, if you have specific recommendations or additional questions, let me know. The main idea of this code is to change representation and simplify where helpful (for example, the long chain of `if` statements to calculate total bond orders were removed). With that in mind, I didn't change the adjacency list format since many unmerged libraries exist for RMG-database. That might be for a future pull request.

rwest commented:
This PR looks good to me. My floating point warning was probably a red herring in this case - sorry for wasting your time - although perhaps raising awareness of it in general is not a bad thing (my modeling class last semester were gobsmacked when I showed them `0.1 + 0.2 != 0.3`).

KEHANG commented:
@goldmanm I think we can merge now. Some final code cleanup thoughts: 

- remove the temporaries commits and 

- I see several identical methods (i.e., `getOrderStr` ) in both `molecule.py` and `group.py`, would be nice if we can consolidate.

goldmanm commented:
The identical methods `getOrderStr`, `getOrderNum`, `setOrderStr`, and `setOrderNum` are object based methods that appear in both `Bond` and `GroupBond` classes. I wrote them separately since they function on different objects which store the `order` attribute differently (`Bond` uses float, `GroupBond` uses list of floats). If you suggest a way to combine methods in this circumstance, I'd be happy to implement it.

KEHANG commented:
@goldmanm it's fine then, please rebase, squash and remove temporary commits.

KEHANG commented:
@goldmanm RMG-database PR is merged, let's wait for the final travis check.
